poh/src/poh_service.rs

```diff
@@ -134,17 +132,23 @@ impl PohService {
                     }
                     Self::tick_producer(
                         poh_recorder,
-                        &poh_exit_,
+                        &poh_exit,
                         ticks_per_slot,
                         hashes_per_batch,
-                        record_receiver,
+                        &record_receiver,
                         Self::target_ns_per_tick(
                             ticks_per_slot,
                             poh_config.target_tick_duration.as_nanos() as u64,
                         ),
                     );
                 }
-                poh_exit_.store(true, Ordering::Relaxed);
+                poh_exit.store(true, Ordering::Relaxed);
+                // Drain record_receiver channel and unblock remaining threads
+                // waiting on the receiver end of the nested channels.
+                for record in record_receiver.try_iter() {
+                    let _ = record.sender.send(Err(PohRecorderError::MaxHeightReached));
+                }
+                drop(record_receiver);
```

Comment:
How would you then call `.try_iter()` then?

Reply:
>  to avoid new items from arriving after the initial try_iter()

There is no "initial" `try_iter`.
The for loop keeps running as long as there are items on the channel even if those items were pushed _after the for loop initially started_.

Also note that there is a
```rust
poh_exit.store(true, Ordering::Relaxed);
```
just before the for loop starts which should stop other validator processes.

`try_iter` returns an iterator which keeps inspecting if the channel is empty or not. As soon as the channel is empty, the iterator returns `None`, the for loop ends, and we drop the receiver side of the channel, which disconnects it from sender and stops sender from sending new items.

_Theoretically_ there is still some minuscule chance that new items are pushed between the time the channel becomes empty and `drop` is invoked. But having run tests for 1000s times, I have not seen that happen once;

Reply:
> try_iter returns an iterator which keeps inspecting if the channel is empty or not. As soon as the channel is empty, the iterator returns None, the for loop ends, and we drop the receiver side of the channel, which disconnects it from sender and stops sender from sending new items.

Yeah these subtle edge cases usually have a way of showing up in tests and other odd behavior. I've also seen times where we run the tests a bunch locally and it seems fine, but then shows up in CI and other people's machines.

I think I would prefer correctness for these subtle edge cases, even if it means introducing more complexity as these are hard to go back and debug/build on top of.